runtime/lua/vim/diagnostic.lua

```diff
@@ -662,15 +662,24 @@ local function get_logical_pos(diagnostic)
 end
 
 --- @param diagnostics vim.Diagnostic[]
+--- @param use_logical_pos boolean
 --- @return table<integer,vim.Diagnostic[]>
-local function diagnostic_lines(diagnostics)
+local function diagnostic_lines(diagnostics, use_logical_pos)
   if not diagnostics then
     return {}
   end
 
   local diagnostics_by_line = {} --- @type table<integer,vim.Diagnostic[]>
   for _, diagnostic in ipairs(diagnostics) do
-    local lnum, _, _, _, valid = get_logical_pos(diagnostic)
+    local lnum ---@type integer
+    local valid ---@type boolean
+
+    if use_logical_pos then
+      lnum, _, _, _, valid = get_logical_pos(diagnostic)
```

Comment:
Unrelated but could we add a comment in `get_logical_pos` explaining what we mean by a diagnostic's logical position?

Reply:
Added a comment